ccmbioinfo/lfs
# CARDS-1623-1724: Add support for DICOM images through the cards:DicomAnswer question type

This PR adds the `cards:DicomAnswer` question type which allows for questionnaires to accept DICOM medical imaging file uploads. Upon DICOM upload, the image is displayed in the Form and the DICOM's metadata is written to the _Notes_ field associated with the question. DICOM file parsing is done on the front-end using the Cornerstone.js library.

It is important to note that at this time, DICOM images that use a _TransferSyntaxUID_ of `1.2.840.10008.1.2.4.70` (JPEG Lossless, Non-hierarchical, First-Order Prediction - Processes 14) are unable to generate an image preview. Metadata extraction from these files still works nonetheless. Using the _dynamic-import_ version of Cornerstone, these previews could be generated, however integrating the _dynamic-import_ version of Cornerstone with CARDS would require upstream modifications to Cornerstone and therefore, for the sake of simplicity, the regular version of the Cornerstone library is used with external error handling for DICOM files with unsupported _TransferSyntaxUID_ values. When these fixes are made to the upstream Cornerstone library, the previously-unsupported _TransferSyntaxUID_ values can simply be removed from the `FORBIDDEN_TRANSFER_SYNTAX_UID` list.

**To test:**

- Build this (`CARDS-1623_1724`) branch with `mvn clean install`
- Start CARDS with the `test` features enabled so that the _DICOM Imaging Test_ Questionnaire is available (`./start_cards.sh --dev --test`).
- Visit `http://localhost:8080` and login as `admin`:`admin`.
- For the sample DICOM files on https://barre.dev/medical/samples/, create a _DICOM Imaging Test_ Form and upload a DICOM file.
- Some of these files (for example, `CT-MONO2-8-abdo.dcm`) will work without modification. That is, an image preview will be displayed and the _Notes_ file will be auto-populated with the DICOM file's metadata.
- Other files (for example `CR-MONO1-10-chest.dcm`) will generate an error message and suggest using `gdcmconv` to fix the file. Follow the suggestion and the converted file upload should work properly.
- The file `CT-MONO2-16-chest.dcm` uses a _TransferSyntaxUID_ of `1.2.840.10008.1.2.4.70`. Uploading this file will display its metadata in the _Notes_ field, but the image preview will be replaced with a message of `Failed to render preview due to Invalid TransferSyntaxUID` as the _TransferSyntaxUID_ (pixel encoding format) is not supported due to upstream issues in the Cornerstone library (see above).

**Regression testing:**

- `FileQuestion` and `Note` have been modified for the implementation of `DicomQuestion`. Test that there are no regressions in these two components.
